src/binding.ts

```diff
@@ -4,61 +4,195 @@
  * sonarlint@sonarsource.com
  * Licensed under the LGPLv3 License. See LICENSE.txt in the project root for license information.
  * ------------------------------------------------------------------------------------------ */
-'use strict';
 
 import * as VSCode from 'vscode';
+import { writeFileSync, mkdirSync, existsSync } from 'fs';
+import { Commands } from './commands';
+import { ConnectionSettingsService } from './settings';
 
 const SONARLINT_CATEGORY = 'sonarlint';
-const PROJECT_BINDING_PROPERTY = 'connectedMode.project';
+const BINDING_SETTINGS = 'connectedMode.project';
+const DEFAULT_CONNECTION_ID = '<default>';
+const OPEN_FOLDER_ACTION = 'Open Folder';
+const BIND_MANUALLY_ACTION = 'Bind Manually';
+
+export interface ProjectBinding {
+    connectionId?: string;
+    serverId?: string;
+    projectKey: string;
+    workspaceFolder: VSCode.WorkspaceFolder;
+}
+
+function createSettingsFileIfNotExists(workspaceFolder: VSCode.WorkspaceFolder) {
+    if (!existsSync(workspaceFolder.uri.fsPath + '/.vscode')) {
+        mkdirSync(workspaceFolder.uri.fsPath + '/.vscode');
+    }
+    try {
+        writeFileSync((workspaceFolder.uri.fsPath + '/.vscode/settings.json'), '', { flag: 'wx' });
+    } catch {
+        return;
+    }
+}
 
 export class BindingService {
 
-  private static _instance: BindingService;
-
-  static init(): void {
-    BindingService._instance = new BindingService();
-  }
-
-  static get instance(): BindingService {
-    return BindingService._instance;
-  }
-
-  async updateBinding(workspaceFolderUri: VSCode.Uri, projectBinding: ProjectBinding): Promise<void> {
-    const config = VSCode.workspace.getConfiguration(SONARLINT_CATEGORY, workspaceFolderUri);
-    return config.update(PROJECT_BINDING_PROPERTY, projectBinding, VSCode.ConfigurationTarget.WorkspaceFolder);
-  }
-
-  async deleteBinding(workspaceFolderUri: VSCode.Uri): Promise<void> {
-    const config = VSCode.workspace.getConfiguration(SONARLINT_CATEGORY, workspaceFolderUri);
-    return config.update(PROJECT_BINDING_PROPERTY, undefined, VSCode.ConfigurationTarget.WorkspaceFolder);
-  }
-
-  async getAllBindings(): Promise<Map<string, Map<string, ProjectBinding[]>>> {
-    const bindings = new Map<string, Map<string, ProjectBinding[]>>();
-    for (const folder of VSCode.workspace.workspaceFolders || []) {
-      const config = VSCode.workspace.getConfiguration(SONARLINT_CATEGORY, folder.uri);
-      const binding = config.get<ProjectBinding>(PROJECT_BINDING_PROPERTY);
-      const projectKey = binding.projectKey;
-      if (projectKey) {
-        const connectionId = binding.connectionId || binding.serverId || '<default>';
-        let connectionBindings = bindings.get(connectionId);
-        if (!bindings.has(connectionId)) {
-          bindings.set(connectionId, new Map<string, ProjectBinding[]>());
+    private static _instance: BindingService;
+
+    static init(getRemoteProjectsForConnection: (connectionId: string) => Thenable<Map<string, string>>,
```

Comment:
I wonder if we should pass the whole client object here (and mock it in the tests - see #265).